packages/asset-swapper/src/utils/market_operation_utils/index.ts

```diff
@@ -240,373 +194,120 @@ export class MarketOperationUtils {
+        // Convert native orders and dex quotes into fill paths.
+        const paths = createFillPaths({
+            side,
+            // Augment native orders with their fillable amounts.
+            orders: createSignedOrdersWithFillableAmounts(side, opts.nativeOrders, opts.orderFillableAmounts),
+            dexQuotes: opts.dexQuotes,
+            targetInput: inputAmount,
+            ethToOutputRate: opts.ethToOutputRate,
+            excludedSources: opts.excludedSources,
+            feeSchedule: opts.feeSchedule,
+        });
+        // Find the optimal path.
+        const optimalPath = findOptimalPath(side, paths, inputAmount, opts.runLimit);
+        if (!optimalPath) {
+            throw new Error(AggregationError.NoOptimalPath);
         }
-    }
-    return optimalPath;
-}
-
-// Gets the total input of a path.
-function getPathInput(path: Fill[]): BigNumber {
-    return BigNumber.sum(...path.map(p => p.input));
-}
-
-// Merges contiguous fills from the same DEX.
-function collapsePath(path: Fill[], isBuy: boolean): CollapsedFill[] {
-    const collapsed: Array<CollapsedFill | NativeCollapsedFill> = [];
-    for (const fill of path) {
-        const makerAssetAmount = isBuy ? fill.input : fill.output;
-        const takerAssetAmount = isBuy ? fill.output : fill.input;
-        const source = (fill.fillData as FillData).source;
-        if (collapsed.length !== 0 && source !== ERC20BridgeSource.Native) {
-            const prevFill = collapsed[collapsed.length - 1];
-            // If the last fill is from the same source, merge them.
-            if (prevFill.source === source) {
-                prevFill.totalMakerAssetAmount = prevFill.totalMakerAssetAmount.plus(makerAssetAmount);
-                prevFill.totalTakerAssetAmount = prevFill.totalTakerAssetAmount.plus(takerAssetAmount);
-                prevFill.subFills.push({ makerAssetAmount, takerAssetAmount });
-                continue;
-            }
+        // Generate a fallback path if native orders are in the optimal paath.
+        let fallbackPath: Fill[] = [];
+        const nativeSubPath = optimalPath.filter(f => f.source === ERC20BridgeSource.Native);
+        if (opts.allowFallback && nativeSubPath.length !== 0) {
+            // The fallback path is, at most, as large as the native path.
+            const fallbackInputAmount = BigNumber.min(inputAmount, getPathSize(nativeSubPath, inputAmount)[0]);
+            fallbackPath =
+                findOptimalPath(side, getFallbackSourcePaths(optimalPath, paths), fallbackInputAmount, opts.runLimit) ||
+                [];
         }
-        collapsed.push({
-            source: fill.fillData.source,
-            totalMakerAssetAmount: makerAssetAmount,
-            totalTakerAssetAmount: takerAssetAmount,
-            subFills: [{ makerAssetAmount, takerAssetAmount }],
-            nativeOrder: (fill.fillData as NativeFillData).order,
+        return createOrdersFromPath([...optimalPath, ...fallbackPath], {
```

Comment:
Still of a mind to have protection here where it's not worth filling this backup path if the price differs from the best path by X amount (aka `slippagePercentage`?)

If that is the case I'd rather eat a revert then slip 10%. Though this is very difficult to simulate at the moment.

I'd be down to fail silently and handle the case if there is no decent fallback path, we tried our best. So the user only gets the optimal path.

Reply:
Added a `maxSlippage` option, but it only applies to the fallback, (reasons discussed offline).